plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java

```diff
@@ -0,0 +1,382 @@
+            }
+            pstmt.close();
+            return caseMap;
+        }
+        catch (SQLException e) {
+            // behavior of base jdbc
+            return ImmutableMap.of();
+        }
+    }
+
+    @Override
+    public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        // this method should ultimately encompass all the expected teradata data types
+        Optional<ColumnMapping> mapping = getForcedMappingToVarchar(typeHandle);
+        if (mapping.isPresent()) {
+            return mapping;
+        }
+
+        switch (typeHandle.jdbcType()) {
+            case Types.TINYINT:
+                return Optional.of(tinyintColumnMapping());
+            case Types.SMALLINT:
+                return Optional.of(smallintColumnMapping());
+            case Types.INTEGER:
+                return Optional.of(integerColumnMapping());
+            case Types.BIGINT:
+                return Optional.of(bigintColumnMapping());
+            case Types.REAL:
+            case Types.DOUBLE:
+            case Types.FLOAT:
+                // teradata float is 64 bit
+                // trino double is 64 bit
+                // teradata float / real / double precision all map to jdbc type float
+                return Optional.of(doubleColumnMapping());
+            case Types.NUMERIC:
+            case Types.DECIMAL:
+                return numberMapping(typeHandle);
+            case Types.CHAR:
+                return Optional.of(charColumnMapping(typeHandle.requiredColumnSize(), deriveCaseSensitivity(typeHandle.caseSensitivity().orElse(null))));
+            case Types.VARCHAR:
+                // see prior note on trino case sensitivity
+                return Optional.of(varcharColumnMapping(typeHandle.requiredColumnSize(), deriveCaseSensitivity(typeHandle.caseSensitivity().orElse(null))));
+            case Types.DATE:
+                return Optional.of(dateColumnMappingUsingLocalDate());
+        }
+
+        if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
+            return mapToUnboundedVarchar(typeHandle);
+        }
+
+        return Optional.empty();
+    }
+
+    private static Optional<ColumnMapping> numberMapping(JdbcTypeHandle typeHandle)
+    {
+        int precision = typeHandle.requiredColumnSize();
+        int scale = typeHandle.requiredDecimalDigits();
+        if (precision > Decimals.MAX_PRECISION) {
+            // this will trigger for number(*) as precision is 40
+            return Optional.of(decimalColumnMapping(createDecimalType(Decimals.MAX_PRECISION, scale)));
+        }
+        return Optional.of(decimalColumnMapping(createDecimalType(precision, scale)));
+    }
+
+    private static ColumnMapping charColumnMapping(int charLength, boolean isCaseSensitive)
+    {
+        if (charLength > CharType.MAX_LENGTH) {
+            return varcharColumnMapping(charLength, isCaseSensitive);
```

Comment:
Fallback to varchar type is weird. Could you leave a code comment?

Reply:
I’ve removed the if loop since it’s not required in this case. Teradata’s maximum CHAR/VARCHAR length (64,000) is always within Trino’s supported CharType.MAX_LENGTH (65,536), so the fallback branch would never be triggered.